xdem/coreg/base.py

```diff
@@ -1154,57 +1363,152 @@ class NotImplementedCoregApply(NotImplementedError):
-    https://peps.python.org/pep-0655/) there is an easy way to specific Required or NotRequired for each key, if we
-    want to change this in the future.
-    """
+class InRandomDict(TypedDict, total=False):
+    """Keys and types of inputs associated with randomization and subsampling."""
 
-    # Common to all coreg objects
+    # Subsample size input by user
     subsample: int | float
-    subsample_final: int
+    # Random state (for subsampling, but also possibly for some fitting methods)
     random_state: int | np.random.Generator | None
 
-    # 1/ Affine metadata
 
-    # Common to all affine transforms
-    centroid: tuple[float, float, float]
-    matrix: NDArrayf
+class OutRandomDict(TypedDict, total=False):
+    """Keys and types of outputs associated with randomization and subsampling."""
 
-    # For translation methods
-    shift_z: np.floating[Any] | float | np.integer[Any] | int
-    shift_x: float
-    shift_y: float
+    # Final subsample size available from valid data
+    subsample_final: int
 
-    # Methods-specific
-    vshift_reduc_func: Callable[[NDArrayf], np.floating[Any]]
 
-    # 2/ BiasCorr classes generic metadata
+class InFitOrBinDict(TypedDict, total=False):
+    """Keys and types of inputs associated with binning and/or fitting."""
+
+    # Whether to fit, bin or bin then fit
+    fit_or_bin: Literal["fit", "bin", "bin_and_fit"]
 
-    # Inputs
-    fit_or_bin: Literal["fit"] | Literal["bin"]
+    # Fit parameters: function to fit and optimizer
     fit_func: Callable[..., NDArrayf]
     fit_optimizer: Callable[..., tuple[NDArrayf, Any]]
+    # Bin parameters: bin sizes, statistic and apply method
     bin_sizes: int | dict[str, int | Iterable[float]]
     bin_statistic: Callable[[NDArrayf], np.floating[Any]]
-    bin_apply_method: Literal["linear"] | Literal["per_bin"]
+    bin_apply_method: Literal["linear", "per_bin"]
+    # Name of variables, and number of dimensions
     bias_var_names: list[str]
     nd: int | None
 
-    # Outputs
+
+class OutFitOrBinDict(TypedDict, total=False):
+    """Keys and types of outputs associated with binning and/or fitting."""
+
+    # Optimized parameters for fitted function, and its error
     fit_params: NDArrayf
     fit_perr: NDArrayf
+    # Binning dataframe
     bin_dataframe: pd.DataFrame
 
-    # Specific inputs or outputs
+
+class InIterativeDict(TypedDict, total=False):
+    """Keys and types of inputs associated with iterative methods."""
+
+    # Maximum number of iterations
+    max_iterations: int
+    # Tolerance at which to stop algorithm (unit specified in method)
+    tolerance: float
+
+
+class OutIterativeDict(TypedDict, total=False):
+    """Keys and types of outputs associated with iterative methods."""
+
+    # Iteration at which the algorithm stopped
+    last_iteration: int
+    # Tolerances of each iteration until threshold
+    all_tolerances: list[float]
+
+
+class InSpecificDict(TypedDict, total=False):
```

Comment:
I'm wondering if all these dictionaries should not live in a separate file. This "base.py" contains a lot of different definitions now.

Reply:
Yes, now that the module is cleaned up, it's much more generic.
I think those sub-dictionaries could remain an exception simply to be more readable (right next to the main `CoregDict`, and I can add this in a xDEM wiki page for implementing new coregistrations!